### PR TITLE
Better validation of `copy_to`.

### PR DESCRIPTION
**Pull request by jpountz**
We are currently quite lenient about the targets of `copy_to`. However in a
number of cases we can detect illegal use of `copy_to` at mapping update time.
For instance, it does not make sense to use object fields as targets of
`copy_to`, or fields that would end up in a different nested document.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 1]
```diff
@@ -648,12 +652,60 @@ private void checkPartitionedIndexConstraints(DocumentMapper newMapper) {
...
     }
 
-    private void checkIndexSortCompatibility(IndexSortConfig sortConfig, boolean hasNested) {
+    private static void checkIndexSortCompatibility(IndexSortConfig sortConfig, boolean hasNested) {
         if (sortConfig.hasIndexSort() && hasNested) {
             throw new IllegalArgumentException("cannot have nested fields when index sort is activated");
         }
     }
 
+    private static void checkCopyToValidity(List<FieldMapper> fieldMappers, Map<String, ObjectMapper> fullPathObjectMappers) {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: Maybe just called it `validateCopyTo`?

File: core/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 2]
```diff
@@ -648,12 +652,60 @@ private void checkPartitionedIndexConstraints(DocumentMapper newMapper) {
...
+        for (FieldMapper mapper : fieldMappers) {
+            if (mapper.copyTo() != null && mapper.copyTo().copyToFields().isEmpty() == false) {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: How could we have CopyTo, but the copy to fields be empty?

File: core/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 3]
```diff
@@ -648,12 +652,60 @@ private void checkPartitionedIndexConstraints(DocumentMapper newMapper) {
...
+                final String sourceScope = getNestedScope(mapper.name(), fullPathObjectMappers);
+                for (String copyTo : mapper.copyTo().copyToFields()) {
+                    if (fullPathObjectMappers.containsKey(copyTo)) {
+                        throw new IllegalArgumentException("Cannot copy to field [" + copyTo + "] since it is mapped as an object");
+                    }
+                    final String targetScope = getNestedScope(copyTo, fullPathObjectMappers);
+                    checkNestedScopeCompatibility(sourceScope, targetScope);
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: Shouldn't this only be called if we have a nested source/target?

**Review by rjernst — APPROVED** [review]
LGTM, I left a couple suggestions

File: core/src/main/java/org/elasticsearch/index/mapper/MapperService.java (hunk 2 above)
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: I'll deal with it in a follow-up PR.

File: core/src/main/java/org/elasticsearch/index/mapper/MapperService.java (hunk 1 above)
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: Will do that.

File: core/src/main/java/org/elasticsearch/index/mapper/MapperService.java (hunk 3 above)
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: I like having `checkNestedScopeCompatibility` entirely responsible for checking what is allowed and what is not.

**Comment by jtreher** [comment]
> For instance, it does not make sense to use object fields as targets of
copy_to, or fields that would end up in a different nested document.

What is the reason that copying from one nested to another should be considered illegal?

**Comment by jpountz** [comment]
Because there may be several of them and there is no way to know which one to copy to.

**Pull request merged**